File: flattenTree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public void flatten(TreeNode root) {
        
        if(root == null){
            return;
        }
        
        TreeNode node = root;
        
        while(node != null){
            if(node.left != null){
                TreeNode rightmost = node.left;
                //find right most element of left sub tree
                while(rightmost.right != null){
                    rightmost = rightmost.right;
                }
                // attach nodes right sub tree to rightmost node of left sub tree
                // change left subtree to right of node and make left null
                rightmost.right = node.right;
                node.right = node.left;
                node.left = null;
            }
            node = node.right;
        }
    }
}
//Time Complexity : O(n)
//Space Complexity: O(1)

File: flattenTree.cpp
// Time Complexity :O(n) n is the number of nodes in the tree
// Space Complexity : O(logn)   
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void flatten(TreeNode* root) {
        helper(root);
        return;
    }
    void helper(TreeNode* root){
        //base
        if (!root) return;
        //logic
        flatten(root->left);
        if (root->left) {
            TreeNode* curr = root->right;
            root->right = root->left;
            root->left = NULL;
            while (root->right) root = root->right;
            root->right = curr;
        }
        flatten(root->right);
    }
};